pkg/backends/calico/routes.go

```diff
@@ -195,24 +197,104 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 	// see if any endpoints are on this node and advertise if so
 	// else remove the route if it also already exists
 	rg.Lock()
-	if rg.advertiseThisService(svc, ep) {
+	defer rg.Unlock()
+
+	advertise := rg.advertiseThisService(svc, ep)
+	if advertise {
 		route := svc.Spec.ClusterIP + "/32"
-		if cur, exists := rg.svcRouteMap[key]; !exists {
+		if cur, exists := rg.svcClusterRouteMap[key]; !exists {
 			// This is a new route - send it through.
-			rg.advertiseRoute(key, route)
+			rg.advertiseClusterRoute(key, route)
 		} else if route != cur {
 			// The route has changed somehow. Send a delete for the old one
 			// and add the new one. We don't expect to hit this, since cluster IPs
 			// are immutable. But, handle it for good measure.
 			logc.Warn("ClusterIP for service changed, adjusting")
-			rg.withdrawRoute(key, cur)
-			rg.advertiseRoute(key, route)
+			rg.withdrawClusterRoute(key, cur)
+			rg.advertiseClusterRoute(key, route)
 		}
-	} else if cur, exists := rg.svcRouteMap[key]; exists {
+	} else if cur, exists := rg.svcClusterRouteMap[key]; exists {
 		// We were advertising this route, but should no longer do so.
-		rg.withdrawRoute(key, cur)
+		rg.withdrawClusterRoute(key, cur)
+	}
+
+	// External IP's
+	if advertise {
+
+		advertisedExternalRoutes := rg.svcExternalRouteMap[key]
+		if advertisedExternalRoutes == nil {
+			advertisedExternalRoutes = make(map[string]bool)
+		}
+
+		externalRoutes := addSuffix(svc.Spec.ExternalIPs, "/32")
+
+		// Advertise any External IP's which we don't already advertise
+		for _, route := range externalRoutes {
+
+			// Only advertise whitelisted External IP's
+			if !rg.isAllowedExternalRoute(route) {
+				log.Infof("External IP not advertised as not whitelisted: %s", route)
+				continue
+			}
+
+			// Advertise route if not already advertised
+			if _, ok := advertisedExternalRoutes[route]; !ok {
+				rg.advertiseExternalRoute(key, route)
```

Comment:
Have you thought at all about what happens if a Service has an ExternalIP that is the same as the ClusterIP of another service?

Or, for that matter, if two services have the same `ExternalIP` associated with them?

Reply:
I think the failure mode here would be when one of the duplicate advertised IP's is removed, as the route for that IP would be deleted, leaving no route advertised for the remaining service.

A mitigation would be to check that before deleting an advertisement, the IP is not present in the cache for any other service.

Reply:
I think if we do the suggestion below to combine these into a single source of routing information, then it simplifies this question a bit.

I'm not actually sure what k8s will do if two services have the same external IPs, or if it is even possible. But, IIUC, since we store the CIDRs to advertise keyed by the service that generated them, we should be OK - even if one withdraws, we'll still advertise the other one.

Reply:
I believe that k8s will allow you to manually assign the same External IP to two different services; assign an External IP to a service which is the same as the service's cluster IP, and assign the same External IP twice to a service.

In all of these scenarios, you would end up in a situation where the same route is "legitimately" being advertised twice from a node.

> even if one withdraws, we'll still advertise the other one

This is not true, because the `routeGenerator client` will discard duplicate route advertisements. So we would only advertise one of the duplicate routes. Then when one of the duplicates is removed, the advertisement would be withdrawn, so we wouldn't advertise the remaining route.

I _think_ this is the case, but I could be wrong.

Reply:
> even if one withdraws, we'll still advertise the other one

This is not true, because the `routeGenerator client` will discard duplicate route advertisements. So we would only advertise one of the duplicate routes. Then when one of the duplicates is removed, the advertisement would be withdrawn, so we wouldn't advertise the remaining route.

I _think_ this is the case, but I could be wrong.

Reply:
Solved this by introducing a `routeAdvertisementCount` map, storing the number of times an advertisement is requested for a route at any one time. We then only withdraw the route if this count is 1 (i.e. nothing else is requesting that this route be advertised).